src/net/server/handlers/message_stream_layer/service.rs

```diff
@@ -0,0 +1,79 @@
+use super::future::ResponseFuture;
+use axum::extract::RequestParts;
+use axum::http::{Request, Response};
+use std::task::{ready, Context, Poll};
+use tokio::sync::mpsc;
+use tokio_util::sync::PollSender;
+use tower::Service;
+
+/// `poll_ready` will first grab a permit via the `PollSender`. Once ready, the sender is injected
+/// into the Request as an Extension. This can only be done once
```

Comment:
I believe that wouldn't be too difficult to backport. Instead of a single `Sender` here, you could have a map of `Sender`s keyed on `ChannelId`, which we will be able to extract from the `Request`

Reply:
we do. we can get the `Request` in `poll_ready`, so we'll have the path

Reply:
no, just checked, and you're right. we only get request in `call`... potentially we could move back to an implementation that uses a `MiddlewareFn` and move the logic into that. we should be ok to do that since we can leave the body unserialized

Reply:
ok, i deleted the entire `message_stream_layer` folder and implemented an async fn middleware. take a look, it's much simpler and gives us access to `Request`

Reply:
nice, so we can inject our global state into the middleware later, when we construct the router.